File: apps/frontend/lib/src/features/notification/domain/entities/notification_entity.dart
import 'package:equatable/equatable.dart';

enum NotificationSeverity {
  info,
  warning,
  error,
  success,
}

enum NotificationChannel {
  inApp,
  email,
  sms,
  push,
  whatsapp,
}

class NotificationEntity extends Equatable {
  const NotificationEntity({
    required this.id,
    this.companyId,
    this.recipientUserId,
    required this.type,
    this.severity = NotificationSeverity.info,
    this.title,
    required this.message,
    this.payload,
    this.isRead = false,
    this.readAt,
    this.channels = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String? companyId;
  final String? recipientUserId;
  final String type;
  final NotificationSeverity severity;
  final String? title;
  final String message;
  final Map<String, dynamic>? payload;
  final bool isRead;
  final DateTime? readAt;
  final List<NotificationChannel> channels;
  final DateTime createdAt;
  final DateTime updatedAt;

  @override
  List<Object?> get props => [
        id,
        companyId,
        recipientUserId,
        type,
        severity,
        title,
        message,
        payload,
        isRead,
        readAt,
        channels,
        createdAt,
        updatedAt,
      ];
}

File: apps/backend/src/modules/notification/entities/notification.entity.ts
import {
  Column,
  Entity,
  Index,
  OneToMany,
} from 'typeorm';
import { TenantBaseEntity } from '../../../shared/database/tenant-base.entity';
import { NotificationChannel } from '../enums/notification-channel.enum';
import { NotificationSeverity } from '../enums/notification-severity.enum';
import { NotificationDelivery } from './notification-delivery.entity';

@Entity({ name: 'notifications' })
@Index(['companyId', 'recipientUserId', 'createdAt'])
export class Notification extends TenantBaseEntity {
  @Column({ type: 'uuid', name: 'recipient_user_id', nullable: true })
  recipientUserId!: string | null;

  @Column({ type: 'varchar', length: 128 })
  type!: string;

  @Column({ type: 'enum', enum: NotificationSeverity, default: NotificationSeverity.INFO })
  severity!: NotificationSeverity;

  @Column({ type: 'varchar', length: 255, nullable: true })
  title!: string | null;

  @Column({ type: 'text' })
  message!: string;

  @Column({ type: 'jsonb', name: 'payload', nullable: true })
  payload!: Record<string, unknown> | null;

  @Column({ type: 'boolean', name: 'is_read', default: false })
  isRead!: boolean;

  @Column({ type: 'timestamp', name: 'read_at', nullable: true })
  readAt!: Date | null;

  @Column({ type: 'enum', enum: NotificationChannel, array: true })
  channels!: NotificationChannel[];

  @OneToMany(
    () => NotificationDelivery,
    (delivery) => delivery.notification,
    { cascade: true },
  )
  deliveries!: NotificationDelivery[];
}
